src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+    }
+    if ((states & io_states::io_failbit) && io_failure()) {
+        throw iostream_system_err(io_failure());
+    }
+}
+
+bool istream::open_nx(const char *path) noexcept
+{
+    fd = bp_sys::open(path, O_RDONLY);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool istream::open(const char *path)
+{
+    bool r = open_nx(path);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::open_nx(const char *path, const int flags) noexcept
+{
+    fd = bp_sys::open(path, O_RDONLY | flags);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool istream::open(const char *path, const int flags)
+{
+    bool r = open_nx(path, flags);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::open_nx(const char *path, const int flags, const mode_t mode) noexcept
+{
+    fd = bp_sys::open(path, O_RDONLY | flags, mode);
+
+    if (fd < 0) {
+        io_error = errno;
+        return false;
+    }
+    return true;
+}
+
+bool istream::open(const char *path, const int flags, const mode_t mode)
+{
+    bool r = open_nx(path, flags, mode);
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+bool istream::close_nx() noexcept
+{
+    if (bp_sys::close(fd) < 0) {
+        io_error = errno;
+        return false; // Failed
+    }
+
+    fd = -1;
+    return true;
+}
+
+bool istream::close()
+{
+    bool r = close_nx();
+    if (!r) {
+        throw_exception_on(io_states::io_failbit);
+    }
+    return r;
+}
+
+int istream::current_state() noexcept
+{
+    int bits = 0;
+    if (eof_state) bits |= io_states::eofbit;
+    if (!buf) bits |= io_states::buffer_failbit;
+    if (string_failed) bits |= io_states::string_failbit;
+    if (io_error) bits |= io_states::io_failbit;
+    return bits;
+}
+
+bool istream::good() noexcept
+{
+    return (current_state() == 0);
+}
+
+bool istream::eof() noexcept
+{
+    return eof_state;
+}
+
+bool istream::buffer_failure() noexcept
+{
+    return (!buf);
+}
+
+bool istream::string_failure() noexcept
+{
+    return string_failed;
+}
+
+int istream::io_failure() noexcept
+{
+    return io_error;
+}
+
+void istream::clear() noexcept
+{
+    eof_state = false;
+    string_failed = false;
+    io_error = 0;
+}
+
+int istream::getc_nx() noexcept
+{
+    if (eof() || buffer_failure() || io_failure()) {
+        // We don't care about string_failure() in getc() function.
+        return -1;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(1);
+        if (r == 0) {
+            eof_state = true;
+            return -1;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    char result = *buf->get_ptr(0);
+    buf->consume(1);
+    return result;
```

Comment:
If the character read is `(char) -1`, this returns -1, which is supposed to indicate failure. You'll need to re-think this API.

I suggest returning `std::pair<char,bool>` (or even a custom struct) instead. It's really bad design to try to wrap a result value and error flag together in a single value.

Reply:
I created `getc_result` class for this purpose:
```c++
// A simple type for keeping result of getc() functions.
class getc_result
{
    public:
    int character;
    bool success;

    getc_result(int character, bool success) noexcept
    {
        this->character = character;
        this->success = success;
    }
};

...

    // Reads and returns one character from file descriptor.
    //
    // Returns: Read character and success of operation as getc_result object.
    //
    // Non _nx variant may throw:
    // dio::iostream_eof on EOF.
    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time).
    // dio::iostream_system_err on system I/O failure.
    getc_result getc_nx() noexcept;
    getc_result getc();
```